clutel/clutel.go

```diff
@@ -31,3 +32,99 @@ func AddToOTELHTTPLabeler(
 		node.AddToOTELHTTPLabeler(ctx),
 	))
 }
+
+// GetSpan retrieves the current OpenTelemetry span from the context.
+func GetSpan(ctx context.Context) trace.Span {
+	return trace.SpanFromContext(ctx)
+}
+
+type spanBuilder struct {
+	kvs  map[string]any
+	opts []trace.SpanStartOption
+}
+
+// NewSpan produces a span builder that allows complete configuration and
+// attribution of the span before it gets started.
+func NewSpan() *spanBuilder {
+	return &spanBuilder{}
+}
+
+// WithAttrs adds attrs to the span.
+func (sb *spanBuilder) WithAttrs(kvs ...any) *spanBuilder {
+	if sb.kvs == nil {
+		sb.kvs = make(map[string]any)
+	}
+
+	// nc.AddValues(stringify.Normalize(kvs...))
+	sb.kvs = stringify.Normalize(kvs...)
+
+	return sb
+}
+
+// WithOpts configures the span with specific otel options.  Note that one
+// of the options can be `trace.WithAttributes()`; don't use that, use
+// spanBuilder.WithAttrs() instead.
+func (sb *spanBuilder) WithOpts(opts ...trace.SpanStartOption) *spanBuilder {
+	if sb.opts == nil {
+		sb.opts = make([]trace.SpanStartOption, 0, len(opts))
+	}
+
+	sb.opts = append(sb.opts, opts...)
+
+	return sb
+}
+
+// Start begins the span with the provided name and attaches it to the context.
+func (sb *spanBuilder) Start(
+	ctx context.Context,
+	name string,
+) context.Context {
+	nc := node.FromCtx(ctx)
+	ctx, spanned := nc.AddSpan(ctx, name, sb.opts...)
+
+	if len(sb.kvs) > 0 {
+		spanned.ID = name
+		spanned = spanned.AddValues(sb.kvs)
+	} else {
+		spanned = spanned.AppendToTree(name)
+	}
+
+	return node.EmbedInCtx(ctx, spanned)
+}
+
+// StartSpan stacks a clues node onto this context and uses the provided
+// name to generate an OTEL span name. StartSpan can be called without
+// adding attributes. Callers should always follow this addition with a
+// closing `defer clues.EndSpan(ctx)`.
+func StartSpan(
```

Comment:
I feel like this should probably just call `spanBuilder.Start` with no options so the logic stays the same in either case

edit: actually, we should probably just stick to one or the other ways of initializing spans

Reply:
Good idea, I'll clean that up.